src/HigherOrderComponent/MapProvider/MapProvider.jsx

```diff
@@ -0,0 +1,97 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import OlMap from 'ol/map';
+
+/**
+ * The MapProvider.
+ *
+ * @type {Object}
+ */
+class MapProvider extends React.Component {
+
+  /**
+   * The properties.
+   * @type {Object}
+   */
+  static propTypes = {
+    /**
+     * The children of this group. Typically a set of `ToggleButton`s.
+     * @type {Object}
+     */
+    children: PropTypes.node,
+
+    /**
+     * The map can be either an OlMap or a Promise that resolves with an OlMap
+     * if your map is created asynchronusly.
+     *
+     * @type {ol.Map|Promise}
+     */
+    map: PropTypes.oneOfType([
+      PropTypes.instanceOf(OlMap),
+      PropTypes.func
+    ])
+  }
+
+  /**
+   * The child context types.
+   * @type {Object}
+   */
+  static childContextTypes = {
+    map: PropTypes.instanceOf(OlMap)
+  }
+
+  /**
+   * The constructor of the MapProvider sets the
+   *
+   * @constructs MapProvider
+   * @param {Object} props The initial props.
+   */
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      map: null,
+      ready: false
+    };
+
+    if (props.map instanceof OlMap) {
+      this.setState({
+        map: props.map,
+        ready: true
+      });
+    } else {
+      props.map().then((map) => {
```

Comment:
Hmm, surprised aboit the invocation of map here. or are we expecting a function that returns a Promise? This might be me not understanding Promises... if so just ignore this.

Reply:
Currently yes. But this makes no sense at all... I gonna check this.